lib/logger/lib/logger.ex

```diff
@@ -695,49 +688,57 @@ defmodule Logger do
           :ok | {:error, :noproc} | {:error, term}
   def bare_log(level, chardata_or_fun, metadata \\ []) do
     case __should_log__(level) do
-      :error -> :ok
-      info -> __do_log__(info, chardata_or_fun, metadata)
+      nil -> :ok
+      level -> __do_log__(level, chardata_or_fun, Map.new(metadata))
     end
   end
 
   @doc false
+  def __should_log__(:warn) do
+    # TODO: Deprecate this one we fully migrate to Erlang levels.
+    # IO.warn("`:warn` level is deprecated, you should use `:warning`")
+    __should_log__(:warning)
+  end
+
   def __should_log__(level) when level in @levels do
-    case __metadata__() do
-      {true, pdict} ->
-        case Logger.Config.log_data(level) do
-          {:discard, _config} -> :error
-          {mode, config} -> {level, mode, config, pdict}
-        end
+    %{level: primary_level} = :logger.get_primary_config()
 
-      {false, _} ->
-        :error
+    if enabled?(self()) and compare_levels(level, primary_level) != :lt do
```

Comment:
I think that `__should_log__` should also be marked as to be removed in future as `logger` will handle all that for us. Only reason for it to stay is the compile-time log removal.

Reply:
I think we will have to keep it because of anonymous functions. We don't want to evaluate them unless we think we will log. Or does Erlang logger allows us to pass anonymous function too? But I think they can't return `:skip` like ours, right?

Reply:
These functions cannot return `:skip` nor `{msg, metadata}` tuple, at least right now. Maybe it could be ported to the Erlang as well.

Reply:
Could we pass an anonymous function and expand it ourselves in a formatter, choosing to potentially skip it? I think we may be able to do it with a global formatter that matches on certain reports.. but that sounds like too much work. Changing OTP would be best.

Reply:
I will try to port it to the OTP as well. For now I do not think that there is proper workaround over it as function itself is expanded before it hits later steps. In theory we could work around it by using report in form of `%{skip: true}` and proper filter, but I do not see any way how to lazily compute the metadata of the log message.

Reply:
Added issues on Erlang JIRA:

- https://bugs.erlang.org/browse/ERL-1085
- https://bugs.erlang.org/browse/ERL-1086